From: Kerin Millar <kerframil@gmail.com>

Add a Hardened Gentoo target to the list of security levels. It's 
designed to provide a comparitively high level of security and to be 
generally suitable for as great a majority of the userbase as possible 
(particularly new users). The patch was originally contributed by Ned 
Ludd <solar@gentoo.org>. This instance was revised by the author with 
contributions from Gordon Malm <bugs-gentoo-org-02@bumpin.org> for the 
hardened-sources-2.6.23-r9 release where it is now a default.

--- a/grsecurity/Kconfig	2008-03-22 17:27:48.000000000 +0000
+++ b/grsecurity/Kconfig	2008-03-22 17:54:44.000000000 +0000
@@ -18,7 +18,7 @@
 choice
 	prompt "Security Level"
 	depends GRKERNSEC
-	default GRKERNSEC_CUSTOM
+	default GRKERNSEC_HARDENED
 
 config GRKERNSEC_LOW
 	bool "Low"
@@ -182,6 +182,96 @@
 	  - Kernel symbol hiding
 	  - Destroy unused shared memory	
 	  - Prevention of memory exhaustion-based exploits
+
+config GRKERNSEC_HARDENED
+	bool "Hardened [Gentoo]"
+	select GRKERNSEC_KMEM
+	select GRKERNSEC_PROC_MEMMAP if (PAX_NOEXEC || PAX_ASLR)
+	select GRKERNSEC_BRUTE
+	select GRKERNSEC_MODSTOP if (MODULES)
+	select GRKERNSEC_HIDESYM
+	select GRKERNSEC_PROC
+	select GRKERNSEC_PROC_USER
+	select GRKERNSEC_PROC_ADD
+	select GRKERNSEC_LINK
+	select GRKERNSEC_FIFO
+	select GRKERNSEC_CHROOT
+	select GRKERNSEC_CHROOT_MOUNT
+	select GRKERNSEC_CHROOT_DOUBLE
+	select GRKERNSEC_CHROOT_PIVOT
+	select GRKERNSEC_CHROOT_CHDIR
+	select GRKERNSEC_CHROOT_CHMOD
+	select GRKERNSEC_CHROOT_FCHDIR
+	select GRKERNSEC_CHROOT_MKNOD
+	select GRKERNSEC_CHROOT_SHMAT
+	select GRKERNSEC_CHROOT_UNIX
+	select GRKERNSEC_CHROOT_FINDTASK
+	select GRKERNSEC_CHROOT_NICE
+	select GRKERNSEC_CHROOT_SYSCTL
+	select GRKERNSEC_CHROOT_CAPS
+	select GRKERNSEC_RESLOG
+	select GRKERNSEC_SIGNAL
+	select GRKERNSEC_FORKFAIL
+	select GRKERNSEC_TIME
+	select GRKERNSEC_PROC_IPADDR
+	select GRKERNSEC_EXECVE
+	select GRKERNSEC_SHM if (SYSVIPC)
+	select GRKERNSEC_DMESG
+	select GRKERNSEC_RANDNET
+	select GRKERNSEC_SYSCTL
+	select GRKERNSEC_SYSCTL_ON
+	select PAX
+	select PAX_EI_PAX
+	select PAX_PT_PAX_FLAGS
+	select PAX_HAVE_ACL_FLAGS
+	select PAX_NOEXEC
+	select PAX_PAGEEXEC
+	select PAX_SEGMEXEC if (X86 && !X86_64)
+	select PAX_EMUTRAMP if (PARISC)
+	select PAX_EMUSIGRT if (PARISC)
+	select PAX_MPROTECT
+	select PAX_ETEXECRELOCS if (ALPHA || IA64 || PARISC)
+	select PAX_EMUPLT if (ALPHA || PARISC || PPC32 || SPARC32 || SPARC64)
+	select PAX_DLRESOLVE if (SPARC32 || SPARC64)
+	select PAX_SYSCALL if (PPC32)
+	select PAX_KERNEXEC if (X86 && !X86_64 && !EFI && !COMPAT_VDSO && !PARAVIRT && X86_WP_WORKS_OK)
+	select PAX_ASLR
+	select PAX_RANDKSTACK if (X86_TSC && !X86_64)
+	select PAX_RANDUSTACK
+	select PAX_RANDMMAP
+	select PAX_MEMORY_SANITIZE
+	select PAX_MEMORY_UDEREF if (X86_32 && !COMPAT_VDSO)
+	help
+	  If you say Y here, a configuration will be used that is endorsed by the
+	  Hardened Gentoo project. It is designed to provide a high level of
+	  security whilst minimizing the chance of incompatibilities with rare
+	  software on your machine. As such, many of the features of grsecurity
+	  and PaX will be enabled. For further information, you should view
+	  <http://grsecurity.net> and <http://pax.grsecurity.net> as well as the
+	  Hardened Gentoo Primer at
+	  <http://gentoo.org/proj/en/hardened/primer.xml>.
+	  
+	  You may wish to emerge paxctl which will allow you to toggle specific
+	  PaX features on problematic binaries. Note that this only works for ELF
+	  binaries that contain a PT_PAX_FLAGS header. In layman's terms this
+	  means that, if you need to toggle PaX features on binaries provided by
+	  applications that are distributed only in binary format (rather than
+	  being built locally from sources), you may need to run paxctl -C on the
+	  binaries beforehand so as to add the missing headers.
+	  
+	  When this level is selected, some options cannot be changed. However,
+	  you may opt to fully customize the options that are selected by choosing
+	  "Custom" in the Security Level menu. You may find it helpful to inherit
+	  the options selected by the "Hardened [Gentoo]" level as a starting
+	  point for further configuration. To accomplish this, select this level
+	  then exit the menuconfig interface, saving changes when prompted. Next,
+	  run make menuconfig again and select the "Custom" level.
+	  
+	  Please note that this security level is not designed to be used in
+	  virtualized environments. If you intend to run the kernel in a virtual
+	  machine then you will probably need to disable the PAX_MEMORY_UDEREF
+	  option in order to avoid an unacceptable impact upon performance.
+
 config GRKERNSEC_CUSTOM
 	bool "Custom"
 	help
